gradle/spotbugs-exclude.xml

```diff
@@ -348,14 +348,11 @@ For a detailed description of spotbugs bug categories, see https://spotbugs.read
     </Match>
 
     <Match>
-        <!-- TODO: fix this (see KAFKA-4996) -->
-        <Or>
-            <Package name="org.apache.kafka.streams.state.internals"/>
+        <!-- Suppress warning about a value that gets initialized
+             before any other threads are created. -->
             <Package name="org.apache.kafka.streams.processor.internals"/>
-            <Package name="org.apache.kafka.streams.processor"/>
-            <Package name="org.apache.kafka.streams"/>
-        </Or>
-        <Bug pattern="IS2_INCONSISTENT_SYNC"/>
+            <Source name="InternalTopologyBuilder.java"/>
+            <Bug pattern="IS2_INCONSISTENT_SYNC"/>
```

Comment:
Interesting... which value is it? Looking at the class, it seems like it's probably the builder methods, which should all be called before passing the pre-built topology to the StreamThreads. I agree this should be safe, but it still might be nice to clean up the class so that this suppression isn't necessary.

Can you expand this XML comment a little to explain what would need to be done to resolve this suppression?

Reply:
It's the `applicationId` -- I guess we should specify in the comment since both you and Boyang asked

Reply:
We get this from the configs, which are passed in to `rewriteTopology` when it's called by `KafkaStreams` constructor. Since we don't always pass the configs in to the topology, we can't initialize this when the InternalTopologyBuilder is constructed.
That said, I'd be in favor of requiring to pass the configs in to the topology _instead_ of passing them in by the `KafkaStreams` constructor. In fact we already force users who want an optimized topology to pass the configs in to the topology AND to the KafkaStreams, which is confusing. If the topology sometimes needs the configs, and we always pass the topology to the KafkaStreams, we should just let KafkaStreams get the configs from the topology.

ie, I'd like to deprecate all `KafkaStreams` constructors with config arguments (which I think is all of them?) and all Topology/StreamsBuilder#build/etc without the config argument. Then we could just initialize the application id in the constructor. Of course, this would require a KIP, and is way out of scope here -- just pointing out why this is so weird

Reply:
Quite a few folks have pushed for this, but I can't help feeling like it's the wrong way to go. It doesn't seem appropriate
to pass the application configs to the topology builder to begin with.

I'm long overdue in writing down my counterproposal, and I won't derail this PR by doing it here. I just wanted to register the possibility that there's a better way out of this mess than forcing users to provide their runtime configurations when compiling their program.

Leaving that aside, I do think you're right. Findbugs is justified in flagging this because, although what we are doing is actually threadsafe, it's hacky and indistingushable from unsafe code.

Reply:
Fair enough. I eagerly await your alternative proposal :P